sdk/python/lib/pulumi/output.py

```diff
@@ -320,50 +322,83 @@ def secret(val: Input[T]) -> 'Output[T]':
+    # https://mypy.readthedocs.io/en/stable/more_types.html#type-checking-the-variants:~:text=All%20of%20the%20arguments%20of%20the,not%20a%20subtype%20of)%20the%20second.
+    def all(*args: Input[T]) -> 'Output[List[T]]':  # type: ignore
+        ...
+
+    @overload
+    @staticmethod
+    def all(**kwargs: Input[T]) -> 'Output[Dict[str, T]]':
+        ...
+
+    @staticmethod
+    def all(*args: Input[T], **kwargs: Input[T]):
         """
-        Produces an Output of Lists from a List of Inputs.
+        Produces an Output of a list (if args i.e a list of inputs are supplied)
+        or dict (if kwargs i.e. keyworded arguments are supplied).
 
         This function can be used to combine multiple, separate Inputs into a single
         Output which can then be used as the target of `apply`. Resource dependencies
         are preserved in the returned Output.
 
+        Examples::
+
+            Output.all(foo, bar) -> Output[[foo, bar]]
+            Output.all(foo=foo, bar=bar) -> Output[{"foo": foo, "bar": bar}]
+
         :param Input[T] args: A list of Inputs to convert.
-        :return: An output of lists, converted from an Input to prompt values.
-        :rtype: Output[List[T]]
+        :param Input[T] kwargs: A list of named Inputs to convert.
+        :return: An output of list or dict, converted from unnamed or named Inputs respectively.
         """
 
         # Three asynchronous helper functions to assist in the implementation:
         # is_known, which returns True if all of the input's values are known,
         # and false if any of them are not known,
-        async def is_known(outputs):
-            is_known_futures = list(map(lambda o: o._is_known, outputs))
+        async def is_known(outputs: Union[dict, list]):
+            if isinstance(outputs, dict):
+                outputs = [outputs[k] for k in outputs]
+            is_known_futures = [o._is_known for o in outputs]
             each_is_known = await asyncio.gather(*is_known_futures)
             return all(each_is_known)
 
         # is_secret, which returns True if any of the input values are secret, and
         # false if none of them are secret.
-        async def is_secret(outputs):
-            is_secret_futures = list(map(lambda o: o._is_secret, outputs))
+        async def is_secret(outputs: Union[dict, list]):
+            if isinstance(outputs, dict):
+                outputs = [outputs[k] for k in outputs]
+            is_secret_futures = [o._is_secret for o in outputs]
             each_is_secret = await asyncio.gather(*is_secret_futures)
             return any(each_is_secret)
 
-        async def get_resources(outputs):
-            resources_futures = list(map(lambda o: o._resources, outputs))
+        async def get_resources(outputs: Union[dict, list]):
+            if isinstance(outputs, dict):
+                outputs = [outputs[k] for k in outputs]
+            resources_futures = [o._resources for o in outputs]
```

Comment:
Kind of a nit - but it does feel like we could avoid forking the implementation of these helpers, and instead pass a list into all of these.  Their behaviour doesn't depend on the dict structure at all - so just pre-computing the list of values once and then calling the existing functions on those would lead to simpler logic throughout I think?